File: LEET-CODE/205.ISOMERIC_STRINGS/ISOMERIC_STRINGS.py
class Solution:
    def isIsomorphic(self, s: str, t: str) -> bool:
        if len(s) != len(t):
            return False  # If the lengths of the strings are different, they cannot be isomorphic

        d = {}  # Dictionary to store mappings from characters in 's' to characters in 't'

        for i in range(len(s)):
            if s[i] not in d:  # If the character from 's' is not yet mapped
                if t[i] in d.values():
                    return False  # If the character from 't' is already mapped to another character in 's', they are not isomorphic
                d[s[i]] = t[i]  # Map the character from 's' to 't'
            else:  # If the character from 's' is already mapped
                if d[s[i]] != t[i]:
                    return False  # If the mapped character from 's' doesn't match the character from 't', they are not isomorphic

        return True  # If all characters are successfully mapped and the mappings match, the strings are isomorphic

                
        

File: LEET-CODE/205.ISOMERIC_STRINGS/ISOMERIC_STRINGS.cpp
#include <bits/stdc++.h>  // Include standard C++ libraries
using namespace std;

class Solution {
public:
    bool isIsomorphic(string s, string t) {
        if (s.size() != t.size()) {
            return false;  // If the lengths of the strings are different, they cannot be isomorphic
        }
        if (s.size() == 0) {
            return true;  // If the strings are empty, they are trivially isomorphic
        }

        map<char, char> m;  // Map to store mappings from characters in 's' to characters in 't'
        map<char, char> m2;  // Map to store mappings from characters in 't' to characters in 's'

        for (int i = 0; i < s.size(); i++) {
            if (m.find(s[i]) == m.end() && m2.find(t[i]) == m2.end()) {
                // If the characters are not mapped in both directions, create new mappings
                m[s[i]] = t[i];  // Map character from 's' to 't'
                m2[t[i]] = s[i];  // Map character from 't' to 's'
            } else {
                // If the characters are already mapped, check if the mappings match
                if (m[s[i]] != t[i] || m2[t[i]] != s[i]) {
                    return false;  // If the mappings don't match, the strings are not isomorphic
                }
            }
        }

        return true;  // If all characters are successfully mapped and the mappings match, the strings are isomorphic
    }
};
